src/api/app/services/workflows/artifacts_collector.rb

```diff
@@ -0,0 +1,49 @@
+module Workflows
+  class ArtifactsCollector
+    def initialize(step:, workflow_run_id:)
+      @step = step
+      @workflow_run_id: workflow_run_id
+    end
+
+    def call
+      # collect artifacts
+      artifacts = case step.type # example: branch_package, configure_repositories, etc... or maybe use the class names like Workflow::Step::ConfigureRepositories, etc...
+                  when 'branch_package'
+                    { step.source_project, step.source_package, step.target_project, step.target_package }
+                  when 'link_package'
+                    { step.source_project, step.source_package, step.target_project, step.target_package }
+                  when 'configure_repositories'
+                    { step.project, step.repositories, step.architectures }
+                  when 'rebuild_package'
+                    { step.project, step.package }
+                  end
```

Comment:
How do you know what is what? By position in the json?

Reply:
For the UI, there will be a view component which knows how to handle every kind of step. With a similar case statement, it will display the artifacts correctly by knowing which step has which keys.

Reply:
I see, this is just pseudo code...